packages/koehnlab/spin_hamiltonians/spin_systems.py

```diff
@@ -0,0 +1,333 @@
+import numpy as np 
+from .phys_const import ge, muNbohr
+from .spin_utils import spinMat
+
+
+numThr = 1e-8
+maxDim = 1000 
+
+class spin:
+
+    def __init__(self,S,type="e",gnuc=None):
+        self.S = S
+        self.type = type
+        self.axes = np.diag([1.,1.,1.])
+        if type=="e":
+            self.g = np.diag([ge,ge,ge])
+        elif type=="n":
+            if gnuc is None:
+                raise Exception("Must provide g factor for nucleus")
+            # the electron g factor is taken here positive, so there is minus sign when
+            # the magnetic moments are built; for the nuclei, we thus use a negative sign
+            # here by default; note that nuclei can have both pos and neg. nuclear g factors
+            self.g = -np.diag([gnuc,gnuc,gnuc])*muNbohr
+        else:
+            raise Exception(f"Unknown type: {type}; allow only 'e' and 'n'")
+        self.ZFaxes = np.diag([1.,1.,1.])
+        self.ZFaxial = 0.
+        self.ZFrhombic = 0.
+        self.ZFten = np.diag([0.,0.,0.])
+
+
+    def set_g(self,g):
+        #if g.shape == (3):
+        self.g = np.matmul(self.axes,np.matmul(np.diag(g),self.axes.T))
+        #elif g.shape == (3,3):
+        #    self.g = g
+        #    # TODO: determine corresp. axis
+        #    # a bit more tricky; do we allow "raw g matrices"? we must sort axes etc.
+        #    raise Exception("spin.set_g: not yet fully implemented")
+        #else:
+        #    raise Exception("spin.set_g: unknown shape of argument; 3 vector or 3x3 matrix expected!")
+
+
+    def set_axes(self,axes):
+        # should we check input? orthog? det=1?
+        axes = np.array(axes)
+        check = np.matmul(axes.T,axes)-np.diag([1.,1.,1.])
+        if np.vdot(check,check) > numThr*numThr:
+            raise Exception("Non-orthogonal axes on input")
+        if np.linalg.det(axes) < 0.:
+            raise Exception("Axes are not defining right-handed system")
+        self.axes = axes
+        self.g = np.matmul(axes,np.matmul(self.g,axes.T))
+
+
+    def set_ZF(self,ZFaxial=0.,ZFrhombic=0.,ZFaxes=None):
+        if ZFaxes is not None:
+            self.ZFaxes = np.array(ZFaxes)
+        else:
+            self.ZFaxes = self.axes
+        self.ZFaxial = ZFaxial
+        self.ZFrhombic = ZFrhombic
+     
+        self.ZFten = np.matmul(self.ZFaxes,
+                     np.matmul(np.diag([-self.ZFaxial/3.+self.ZFrhombic,-self.ZFaxial/3.-self.ZFrhombic,2*self.ZFaxial/3.]),
+                               self.ZFaxes.T))
+
+
+    def getSpinMat(self):
```

Comment:
Inconsistent naming of functions - other functions use snake_case whereas here you use camelCase.

Reply:
I am probably very inconsistent at times ... I have to have a look at all the other functions, do not remember ...

Reply:
OK, I will probably introduce snake consistently. I could do the renaming of the class then as well.